### PR TITLE
Support AWS Elastic Network Interfaces (ENI)

### PR DESCRIPTION
**Pull request by peterbeams**
Adding a new resource called aws_network_interface to create network interfaces and attach them to instances.

Mentioned in these issues ...
[Add support for ENI's to `aws_instance` resource #84] (https://github.com/hashicorp/terraform/issues/84)
[AWS Provider Coverage #28] (https://github.com/hashicorp/terraform/issues/28) 

Sample:
```
resource "aws_vpc" "foo" {
    cidr_block = "172.16.0.0/16"    
}

resource "aws_subnet" "foo" {
    vpc_id = "${aws_vpc.foo.id}"
    cidr_block = "172.16.10.0/24"
    availability_zone = "us-west-2a"
}

resource "aws_subnet" "bar" {
    vpc_id = "${aws_vpc.foo.id}"
    cidr_block = "172.16.11.0/24"
    availability_zone = "us-west-2a"
}

resource "aws_security_group" "foo" {
  vpc_id = "${aws_vpc.foo.id}"
  description = "foo"
  name = "foo"  
}

resource "aws_instance" "foo" {
	ami = "ami-c5eabbf5"
	instance_type = "t2.micro"
	subnet_id = "${aws_subnet.bar.id}"
	associate_public_ip_address = false
	private_ip = "172.16.11.50"
}

resource "aws_network_interface" "bar" {
    subnet_id = "${aws_subnet.foo.id}"   
    private_ips = ["172.16.10.100"]
    security_groups = ["${aws_security_group.foo.id}"]    
    attachment {
    	instance = "${aws_instance.foo.id}"
    	device_index = 1
    }
    tags {
        Name = "bar_interface"
    }
}
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_network_interface.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+	}
+	
+	log.Printf("[DEBUG] Creating network interface")
+	resp, err := ec2conn.CreateNetworkInterface(request)
+	if err != nil {
+		return fmt.Errorf("Error creating ENI: %s", err)
+	}
+
+	new_interface_id := *resp.NetworkInterface.NetworkInterfaceID
+	d.SetId(new_interface_id)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_network_interface.go]: small nitpick in that `new_interface_id` is never used beyond setting here, so `d.SetId(*resp.NetworkInterface.NetworkInterfaceID)` works just as well. Don't feel you _need_ to change this though :smile:

File: builtin/providers/aws/resource_aws_network_interface.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	if err != nil {
+		if ec2err, ok := err.(aws.APIError); ok && ec2err.Code == "InvalidNetworkInterfaceID.NotFound" {
+			// The ENI is gone now, so just remove it from the state
+			d.SetId("")
+			return nil
+		}
+		
+		return fmt.Errorf("Error retrieving ENI: %s", err)
+	}
+	if len(describeResp.NetworkInterfaces) != 1 {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_network_interface.go]: The [ENI documentation][1] states "You can create and attach additional network interfaces.". 
Is that allowed here? Seems you should be looking for the network interface with a matching ID?

[1]: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html

Comment (catsby) [inline builtin/providers/aws/resource_aws_network_interface.go]: Re-reading, we're calling `DescribeNetworkInterfacesRequest` with a specific `id`, so this should return 1 or 0, right? In which case, ignore me :smile:

File: builtin/providers/aws/resource_aws_network_interface.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+func convertToGroupIds(dtos []ec2.GroupIdentifier) []string {
+	ids := make([]string, 0, len(dtos))
+	for _, v := range dtos {
+		group_id := *v.GroupID
+		ids = append(ids, group_id)
+	}
+	return ids
+}
+
+func convertToPrivateIPAddresses(ips []interface{}) []ec2.PrivateIPAddressSpecification {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_network_interface.go]: the convention we have for your conversion type methods is `flatten<things>` and `expand<things>`. You can see examples in [builtin/providers/aws/structure.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/structure.go) 

If possible, these should be unit tested as well, similar to [structure_test.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/structure_test.go)

**Comment by catsby** [comment]
This looks very promising, thanks! 
I had some nitpicks, questions, and requests, but other than that it looks solid so far. 
Could you please take a look at my feedback, and also run `go fmt` on the files?

**Comment by peterbeams** [comment]
Thanks for the feedback, will take a look through and make some changes.

**Comment by jwestboston** [comment]
Very excited to have this feature.

**Comment by catsby** [comment]
Hey @peterbeams do you think you'll be able to complete the work here? 
I can try and complete it if not. 
Thanks again!

**Comment by peterbeams** [comment]
Hi @catsby - I'm working on the tweaks now - should have it pushed in couple of hours.  Sorry for delay - got carried away with using TF to build my environment!

File: builtin/providers/aws/resource_aws_network_interface.go (hunk 1 above)
Comment (peterbeams) [inline builtin/providers/aws/resource_aws_network_interface.go]: Sorted this one with new commit

File: builtin/providers/aws/resource_aws_network_interface.go (hunk 2 above)
Comment (peterbeams) [inline builtin/providers/aws/resource_aws_network_interface.go]: Yes, looking for just the one specific id of the adapter we're describing.  Multiple adapters can be attached, but they'll each be a different resource.

**Comment by peterbeams** [comment]
Done with the updates, hopefully up to scratch now.  Feel free to made any further tweaks you like though.  Thanks!

**Comment by catsby** [comment]
LGTM! 
Tags aren't fully supported though, but I can add those :smile:

**Comment by catsby** [comment]
Finished `tag` support in a063ebe992 , thanks again!

**Comment by peterbeams** [comment]
Great, thanks @catsby.

**Comment by mdallman2570** [comment]
I was looking through the code, but I did not see the ability to add existing ENIs to the instance.  Is there an easy way to attach a newly generated instance to an existing ENI?  Support under the aws_instance block would be huge.  I didn't see it present, but that doesn't mean it isn't there @peterbeams

**Comment by ajmath** [comment]
I'm with @mdallman2570.  We can attach the ENI to the instance after it's been created but I don't see a way to associate the ENI to the instance when it's creating the instance.

This is useful for me to be able to have the ENI be assigned to eth0 on the instance and avoid any headache of updating the networking on the instance to use the ENI.

**Comment by catsby** [comment]
@ajmath I believe https://github.com/hashicorp/terraform/issues/2998 is requesting the same thing, yeah?

**Comment by ajmath** [comment]
@catsby yes it is.  Thanks

**Comment by catsby** [comment]
Great, just making sure I understood. I can't promise when I'll get to it, but at least it's being tracked separately

**Comment by automaticgiant** [comment]
There is code in [`resource_aws_instance.go`](../blob/master/builtin/providers/aws/resource_aws_instance.go#L356) to do this, is there not?
Though you have to have a an identifier for the eni that `aws_instance` will take.
I guess you can get it, according to [`resource_aws_network_interface.go'](../blob/master/builtin/providers/aws/resource_aws_network_interface.go#L119).
I'll test it. This is important.

**Pull request merged**